src/npctalk.cpp

```diff
@@ -3218,84 +3221,82 @@ std::string give_item_to( npc &p, bool allow_use, bool allow_carry )
         // Eating first, to avoid evaluating bread as a weapon
-        const auto consume_res = try_consume( p, given, no_consume_reason );
-        if( consume_res == CONSUMED_ALL ) {
-            g->u.i_rem( &given );
-        }
+        const auto consume_res = try_consume( p, given, reason );
         if( consume_res != REFUSED ) {
+            if( consume_res == CONSUMED_ALL ) {
+                g->u.i_rem( &given );
+            }
             g->u.moves -= 100;
             if( given.is_container() ) {
                 given.on_contents_changed();
             }
-            return _( "Here we go…" );
-        }
-    }
-
-    bool taken = false;
-    int our_ammo = p.ammo_count_for( p.weapon );
-    int new_ammo = p.ammo_count_for( given );
-    const double new_weapon_value = p.weapon_value( given, new_ammo );
-    const double cur_weapon_value = p.weapon_value( p.weapon, our_ammo );
-    if( allow_use ) {
-        add_msg( m_debug, "NPC evaluates own %s (%d ammo): %0.1f",
-                 p.weapon.type->get_id(), our_ammo, cur_weapon_value );
-        add_msg( m_debug, "NPC evaluates your %s (%d ammo): %0.1f",
-                 given.type->get_id(), new_ammo, new_weapon_value );
-        if( new_weapon_value > cur_weapon_value ) {
+        }// wield it if its a weapon
+        else if( new_weapon_value > cur_weapon_value ) {
             p.wield( given );
+            reason = _( "Thanks, I'll wield that now." );
             taken = true;
-        }
-
-        // HACK: is_gun here is a hack to prevent NPCs wearing guns if they don't want to use them
-        if( !taken && !given.is_gun() && p.wear_if_wanted( given ) ) {
+        }// HACK: is_gun here is a hack to prevent NPCs wearing guns if they don't want to use them
+        else if( !given.is_gun() && given.is_armor() ) {
+            //if it is impossible to wear return why
+            ret_val<bool> can_wear = p.can_wear( given, true );
+            if( !can_wear.success() ) {
+                reason = can_wear.str();
+            } else {
+                //if we can wear it with equip changes prompt first
+                can_wear = p.can_wear( given );
+                if( ( can_wear.success() ||
+                      query_yn( std::string( can_wear.c_str() ) += std::string(
+                                    _( " Should I take something off?" ) ) ) )
+                    && p.wear_if_wanted( given, reason ) ) {
+                    taken = true;
+                } else {
+                    reason = can_wear.c_str();
+                }
+            }
+        } else {
+            reason += _( "My current weapon is better than this." );
+            reason += "\n" + string_format( _( "(new weapon value: %.1f vs %.1f)." ), new_weapon_value,
+                                            cur_weapon_value );
+        }
+    } else {//allow_use is false so try to carry instead
+        if( p.can_pickVolume( given ) &&
+            p.can_pickWeight( given ) ) {
```

Comment:
```suggestion
        if( p.can_pickVolume( given ) && p.can_pickWeight( given ) ) {
```
Certainly this fits on one line.